File: scripts/threadpaper.rb
#!/usr/bin/env ruby
#
#  threadpaper -- make wallpaper from a Threadless sample image
#
#  by Rich Lafferty <rich@lafferty.ca> -- released into the public
#  domain on December 21, 2005.
#
#  default filename changed to shirt-title - joshua@cpan.org 23 Oct 2006
#
#  ported to ruby - ymendel@pobox.com 13 Jul 2009

require 'rubygems'
require 'rmagick'
require 'optparse'
require 'net/http'

# use strict;
# use LWP::Simple qw(getstore);
# use Image::Magick;
# use Getopt::Long;
# use File::Basename;
# # use Smart::Comments;

logo = "./threadless-logo.png";
tmp = "/tmp";

# Image::Magick::Histogram's returned color values are the true color
# multiplied by this magic number:
hist_magic = 257;

ENV['PATH'] = '/usr/bin:/bin:/usr/local/bin';
ENV['BASH_ENV'] = '';


#------------------------------------------------------------------------
# get and verify arguments

OPTIONS = { :width => 1440, :height => 900, :scale => 100, :place => 'SouthEast' }

parser = OptionParser.new do |opts|
  opts.banner = <<BANNER
Usage: #{File.basename($0)} [options] [threadless-url]

Options are:
BANNER

  opts.separator ''
  opts.on('-w', '--width=WIDTH',
          "width of wallpaper (default: #{OPTIONS[:width]})")   { |width|   OPTIONS[:width] = width }
  opts.on('-h', '--height=HEIGHT',
          "height of wallpaper (default: #{OPTIONS[:height]})") { |height|  OPTIONS[:height] = height }
  opts.on('-c', '--center',
          'center image instead of lower corner') { |center|  OPTIONS[:place] = 'Center' }
  opts.on('-l', '--logo',
          'include color-matched Threadless logo') { |logo|  OPTIONS[:logo] = true }
  opts.on('-s', '--scale=PERCENT',
          'scale image by PERCENT') { |pct|  OPTIONS[:scale] = pct }
  opts.on('-f', '--file=FILENAME',
          'output filename', '(default: <shirt-title in lowercase>.png)') { |file|  OPTIONS[:file] = file }
  opts.on('--help',
          'Show this help message.') { puts opts; exit }

  opts.parse!(ARGV)
end

histfile = File.join(ENV['HOME'], '.threadpaper_history')
# write_history();

url = ARGV[0]
md = url.to_s.match(Regexp.new('^http://.*threadless.*/(\d+)/([^/]+)'))

unless md
  puts parser
  exit
end

num   = md[1]
title = md[2]

OPTIONS[:file] ||= "#{title}.png".downcase

[:width, :height].each do |dim|
  begin
    OPTIONS[dim] = Integer(OPTIONS[dim])
  rescue
    puts parser
    exit
  end
end
geometry = OPTIONS.values_at(:width, :height).collect { |x|  "#{x}!" }.join('x')

tmpfile = File.join(tmp, "th_orig_#{$$}.gif")
realurl = "http://media.threadless.com/product/#{num}/zoom.gif";

File.open(tmpfile, 'w') do |f|
  f.print Net::HTTP.get(URI.parse(realurl))
end

orig = Magick::Image.read(tmpfile).first
File.unlink(tmpfile)
orig.resize!(OPTIONS[:scale]/100.0) unless OPTIONS[:scale] == 100

bg = orig.clone
bg.crop!(0,0,1,1)
bg.resize!(OPTIONS[:width], OPTIONS[:height])
bg.composite!(orig, Magick.const_get("#{OPTIONS[:place]}Gravity"), Magick::CopyCompositeOp)

bg.write(OPTIONS[:file])

exit

# # add logo
# 
# if ($do_logo)
# {
#   
#   ### Adding logo
#   
#     my $lo = Image::Magick->new;
#     $lo->read($logo);
# 
#     my @bghist   = gethist($bg);
#     my $bgcolor = hist2rgb($bghist[0]);
# 
#     my @orighist = gethist($orig);
#     my $color0 = hist2rgb($orighist[0]);
#     my $color1 = hist2rgb($orighist[1]);
# 
#     my $logocolor = $bgcolor eq $color0 ? $color1 : $color0;
# 
#     $lo->Opaque( color => "rgb(0,0,0)", fill => $logocolor );
#     $lo->Resize( geometry => "100x" );
#     $bg->Composite( image => $lo, gravity => "southeast");
# }
# 
# sub gethist
# {
#     my $im = shift;
#     my @h = $im->Histogram();
#     
#     # The returned values are an array of red, green, blue, opacity, and
#     # count values [all shoved into one long array].
#     my @hist;
# 
#     push @hist, [splice(@h, 0, 5)] while @h;
#     
#     return reverse sort { $a->[4] <=> $b->[4] } @hist;    
# }
# 
# sub hist2rgb
# {
#     my $hist = shift;
#     return sprintf("rgb(%d,%d,%d)", $hist->[0] / $hist_magic,
#                                     $hist->[1] / $hist_magic,
#                                     $hist->[2] / $hist_magic);
# }
#
# sub write_history
# {
#   my $binary = basename $0;
#   
#   open  LOG, ">> $histfile" or die "Can't open $histfile: $!";
#   print LOG "$binary @ARGV\n";
#   close LOG;
# }

File: bin/threadpaper.pl
#!/usr/bin/env perl -wT
#
#  threadpaper -- make wallpaper from a Threadless sample image
#
#  by Rich Lafferty <rich@lafferty.ca> -- released into the public
#  domain on December 21, 2005.
#
#  default filename changed to shirt-title - joshua@cpan.org 23 Oct 2006
#

use strict;
use LWP::Simple qw(getstore);
use Image::Magick;
use Getopt::Long;
use File::Basename;
# use Smart::Comments;

my $logo = "./threadless-logo.png";
my $tmp = "/tmp";

# Image::Magick::Histogram's returned color values are the true color
# multiplied by this magic number:
my $hist_magic = 257;

$ENV{PATH} = '/usr/bin:/bin:/usr/local/bin';
$ENV{BASH_ENV} = '';


#------------------------------------------------------------------------
# get and verify arguments

my $width = 1920;
my $height = 1200;
my $center;
my $do_logo;
my $scale = 100;
my $outfile;
my $histfile = "$ENV{HOME}/.threadpaper_history";

($histfile) = $histfile =~ m{^([/\w+.]+)$}; # detaint
write_history();

GetOptions ("width=i"   => \$width,
            "height=i"  => \$height,
            "center"    => \$center,
            "centre"    => \$center,
            "logo"      => \$do_logo,
            "scale=i"   => \$scale,
            "file=s"    => \$outfile ) or die usage();

my $type = $center ? 'center' : 'southeast';

usage() unless @ARGV;
my $url = $ARGV[0];

usage() unless $url =~ m|^http://.*threadless.*/(\d+)/([^/]+)|;

my $num = $1;
my $title = $2;

$outfile = "\L$title.png" unless $outfile;

usage() unless $width =~ /^(\d+)$/; 
$width = $1;
usage() unless $height =~ /^(\d+)$/;
$height = $1;
my $geometry = "${width}!x${height}!";

#------------------------------------------------------------------------
# retrieve image from Threadless

my $file = "$tmp/th_orig_$$.gif";
my $realurl = "http://media.threadless.com/product/$num/zoom.gif";
getstore($realurl, $file) or die "Can't get $url\n"; ### Downloading

my $orig = Image::Magick->new;
$orig->Read($file);
unlink($file);

$orig->Resize( geometry => "${scale}%" ) if $scale != 100; ### Resizing

#------------------------------------------------------------------------
# prepare background

### Building new background

my $bg = $orig->Clone();
$bg->Crop(   geometry => '1x1+0+0' );
$bg->Resize( geometry => $geometry );
$bg->Composite( image => $orig, gravity => $type );

#------------------------------------------------------------------------
# add logo

if ($do_logo)
{
	
	### Adding logo
	
    my $lo = Image::Magick->new;
    $lo->read($logo);

    my @bghist   = gethist($bg);
    my $bgcolor = hist2rgb($bghist[0]);

    my @orighist = gethist($orig);
    my $color0 = hist2rgb($orighist[0]);
    my $color1 = hist2rgb($orighist[1]);

    my $logocolor = $bgcolor eq $color0 ? $color1 : $color0;

    $lo->Opaque( color => "rgb(0,0,0)", fill => $logocolor );
    $lo->Resize( geometry => "100x" );
    $bg->Composite( image => $lo, gravity => "southeast");
}

#------------------------------------------------------------------------
# finish

$bg->Write($outfile); ### Writing file
exit 0;

#------------------------------------------------------------------------
# Subs

sub gethist
{
    my $im = shift;
    my @h = $im->Histogram();
    
    # The returned values are an array of red, green, blue, opacity, and
    # count values [all shoved into one long array].
    my @hist;

    push @hist, [splice(@h, 0, 5)] while @h;
    
    return reverse sort { $a->[4] <=> $b->[4] } @hist;    
}

sub hist2rgb
{
    my $hist = shift;
    return sprintf("rgb(%d,%d,%d)", $hist->[0] / $hist_magic,
                                    $hist->[1] / $hist_magic,
                                    $hist->[2] / $hist_magic);
}

sub usage
{
    warn "Usage: $0 [options] [threadless-url]\n";
    warn "\n";
    warn "\t--width WIDTH    width of wallpaper (default: $width)\n";
    warn "\t--height HEIGHT  height of wallpaper (default: $height)\n";
    warn "\t--center         center image instead of lower corner\n";
    warn "\t--logo           include color-matched Threadless logo\n";
    warn "\t--scale PERCENT  scale image by PERCENT\n";
    warn "\t--file FILENAME  output filename\n";
    die "\t                      (default: <shirt-title in lowercase>.png)\n";
}

sub write_history
{
	my $binary = basename $0;
	
	open  LOG, ">> $histfile" or die "Can't open $histfile: $!";
	print LOG "$binary @ARGV\n";
	close LOG;
}
